akka-typed/src/main/scala/akka/typed/scaladsl/Actor.scala

```diff
@@ -193,6 +192,33 @@ object Actor {
   def onSignal[T](handler: PartialFunction[(ActorContext[T], Signal), Behavior[T]]): Behavior[T] =
     immutable[T]((_, _) ⇒ same).onSignal(handler)
 
+  /**
+   * Behavior that stops all child actors, waits for them to stop and then
+   * runs some user defined cleanup before stopping itself.
+   */
+  def gracefulStop[T](cleanup: ActorContext[T] ⇒ Unit): Behavior[T] =
+    Actor.deferred { context ⇒
+      if (context.children.isEmpty) {
+        cleanup(context)
+        Behavior.stopped
+      } else {
+        context.children.foreach { cld ⇒
+          context.watch(cld)
+          context.stop(cld)
+        }
+      }
+      onSignal {
+        case (ctx, Terminated(_)) ⇒
```

Comment:
use a guard on this to get one case where we stop and one where we return `same`?

Reply:
yes